---
title: "Feature Selection: Genotype Data"
format: pdf
editor: visual
bibliography: references.bib
editor_options:
  markdown:
    wrap: 72
---

The maize GxE [competition
data](https://drive.google.com/drive/folders/1leYJY4bA3341S-JxjBIgmmAWMwVDHYRb)
includes genotype data for 4,928 individuals at 437,214 variant sites.
To reduce the number of genetic features in this dataset, we leveraged
knowledge of maize evolutionarily history to identify a subset of SNPs
important for adaptation to diverse environments. Specifically, we
select a subset of 2,513 SNPs that were previously implicated in
adaptation of 4,471 maize landraces from across Mexico, Central America,
and South America to altitude [@romeronavarro2017]. The majority of
these SNPs (61%) were also associated with variation in flowering time
in experiments [@romeronavarro2017]. Early flowering is an important
adaptation to high latitudes and high elevations, where there is a
shorter growing season. SNPs predictive of flowering time variation
could be important for explaining variation in performance of different
individuals across the G2F locations.

Genotype data for the landraces studied by @romeronavarro2017 were
downloaded from the CIMMYT Seeds of Discovery project [@hearne2012].
Hapmap files were subset to SNPs present in Supplemental Table 6 from
@romeronavarro2017, combined, and converted to VCF using TASSEL version
5.2.64 [@bradbury2007].

```{bash}
# filter 
~/tools/tassel-5-standalone/run_pipeline.pl -Xmx30g -fork1 -h ~/scratch/maize/CIMMYT/GBS/AllZeaGBSv2.7_SEED_Beagle4_chr9.hmp.txt -includeSiteNamesInFile Romero_altitude.txt -export chr9.alt.flt

# combine
~/tools/tassel-5-standalone/run_pipeline.pl -Xmx30g -fork1 -h alt.hmp.txt -fork2 -h alt2.hmp.txt -fork3 -h alt3.hmp.txt -fork4 -h chr10.alt.flt.hmp.txt -combine5 -input1 -input2 -input3 -input4 -mergeGenotypeTables -export romero_alt_zea27 -runfork1 -runfork2 -runfork3 -runfork4

# convert to VCF 
~/tools/tassel-5-standalone/run_pipeline.pl -Xmx30g -fork1 -h romero_alt_zea27.hmp.txt -export romero_alt_zea27 -exportType VCF
```

Coordinates based on the AGPv2 reference genome were then converted to
Zm-B73-REFERENCE-NAM-5.0Zm using the [Assembly Converter online
tool](https://plants.ensembl.org/Zea_mays/Tools/AssemblyConverter?db=core)
from EnsemblPlants, which is based on CrossMap [@zhao2014].

## References
